File: apps/backend/src/routes/health.py
"""
Health check routes for the JurisAI API with blue-green deployment support.
"""

import logging
import platform
import time
from typing import Any, Dict

import psutil
from fastapi import APIRouter, Depends, FastAPI, Response
from sqlalchemy.orm import Session

from src.core.database import get_db
from src.core.health import health_checker

router = APIRouter(prefix="/health", tags=["health"])

# Start time for uptime calculation
START_TIME = time.time()


@router.get("/")
async def health_check() -> Dict[str, Any]:
    """
    Basic health check endpoint for blue-green deployments.

    Returns:
        Dict[str, Any]: Health status
    """
    return {"status": "healthy", "uptime": f"{int(time.time() - START_TIME)} seconds"}


@router.get("/ready")
async def readiness_check(response: Response) -> Dict[str, Any]:
    """
    Readiness check for blue-green traffic routing.
    
    Returns:
        Dict[str, Any]: Readiness status
    """
    readiness_status = await health_checker.get_readiness_status()
    
    if not readiness_status["ready"]:
        response.status_code = 503
        
    return readiness_status


@router.get("/live")
async def liveness_check(response: Response) -> Dict[str, Any]:
    """
    Liveness check for blue-green deployment monitoring.
    
    Returns:
        Dict[str, Any]: Liveness status
    """
    health_status = await health_checker.get_comprehensive_health()
    
    if health_status["status"] == "unhealthy":
        response.status_code = 503
        
    return health_status


@router.get("/system")
async def system_info() -> Dict[str, Any]:
    """
    Get detailed system information.

    Returns:
        Dict[str, Any]: System information
    """
    try:
        return {
            "os": platform.system(),
            "os_version": platform.version(),
            "python_version": platform.python_version(),
            "cpu_usage_percent": psutil.cpu_percent(),
            "memory_usage_percent": psutil.virtual_memory().percent,
            "disk_usage_percent": psutil.disk_usage("/").percent,
        }
    except Exception as e:
        logging.error(f"Error getting system info: {e}")
        return {
            "status": "error",
            "message": "Could not retrieve system information",
            "os": platform.system(),
            "python_version": platform.python_version(),
        }


@router.get("/ai-models")
async def ai_models_check() -> Dict[str, Any]:
    """
    Check AI models availability and status.

    Returns:
        Dict[str, Any]: AI models status
    """
    models_status = {
        "rag_available": False,
        "summarizer_available": False,
        "document_processor_available": False,
    }

    # Check RAG model
    try:
        from src.routes.search import RAG_AVAILABLE

        models_status["rag_available"] = RAG_AVAILABLE
    except ImportError:
        pass

    # Check summarizer
    try:
        from libs.ai_models.src.summarization.summarizer import LegalDocumentSummarizer

        models_status["summarizer_available"] = True
    except ImportError:
        pass

    # Check document processor
    try:
        from libs.ai_models.src.document_processing.processor import DocumentProcessor

        models_status["document_processor_available"] = True
    except ImportError:
        pass

    return {
        "status": "operational" if any(models_status.values()) else "limited",
        "models": models_status,
    }


@router.get("/database")
async def database_check(db: Session = Depends(get_db)) -> Dict[str, Any]:
    """
    Check database connection status.

    Args:
        db (Session): Database session

    Returns:
        Dict[str, Any]: Database connection status
    """
    try:
        # Try a simple query to verify database connection
        db.execute("SELECT 1")
        return {
            "status": "connected",
            "type": db.bind.dialect.name,
            "message": "Database connection successful"
        }
    except Exception as e:
        logging.error(f"Database connection error: {e}")
        return {
            "status": "error",
            "message": f"Database connection failed: {str(e)}"
        }


@router.get("/full")
async def full_health_check(
    db: Session = Depends(get_db)
) -> Dict[str, Any]:
    """
    Comprehensive health check of all system components for blue-green deployments.

    Args:
        db (Session): Database session

    Returns:
        Dict[str, Any]: Complete system health status
    """
    # Use the enhanced health checker
    comprehensive_health = await health_checker.get_comprehensive_health()
    
    # Get additional legacy info for backward compatibility
    sys_info = await system_info()
    db_status = await database_check(db)
    ai_status = await ai_models_check()
    
    # Merge with comprehensive health data
    comprehensive_health.update({
        "legacy_system": sys_info,
        "legacy_database": db_status,
        "legacy_ai_models": ai_status
    })
    
    return comprehensive_health

File: apps/frontend/lib/api/health.ts
/**
 * Health API module for JurisAI
 * Provides functions for checking system health and component status
 */
import apiClient, { ApiResponse } from './client';

// Types for health check responses
export interface BasicHealthResponse {
  status: string;
  uptime: string;
}

export interface SystemInfoResponse {
  os: string;
  os_version: string;
  python_version: string;
  cpu_usage_percent: number;
  memory_usage_percent: number;
  disk_usage_percent: number;
}

export interface DatabaseStatusResponse {
  status: string;
  type: string;
  message: string;
}

export interface AIModelsResponse {
  status: string;
  models: {
    rag_available: boolean;
    summarizer_available: boolean;
    document_processor_available: boolean;
  };
}

export interface FullHealthResponse {
  status: string;
  timestamp: number;
  uptime: string;
  database: DatabaseStatusResponse;
  system: SystemInfoResponse;
  ai_models: AIModelsResponse;
  version: string;
}

/**
 * Get basic health status
 * @returns Promise with basic health status
 */
export const getBasicHealth = async (): Promise<BasicHealthResponse> => {
  try {
    const response = await apiClient.get<BasicHealthResponse>('/health');
    return response.data;
  } catch (error) {
    throw error;
  }
};

/**
 * Get system information
 * @returns Promise with system information
 */
export const getSystemInfo = async (): Promise<SystemInfoResponse> => {
  try {
    const response = await apiClient.get<SystemInfoResponse>('/health/system');
    return response.data;
  } catch (error) {
    throw error;
  }
};

/**
 * Get database status
 * @returns Promise with database connection status
 */
export const getDatabaseStatus = async (): Promise<DatabaseStatusResponse> => {
  try {
    const response = await apiClient.get<DatabaseStatusResponse>('/health/database');
    return response.data;
  } catch (error) {
    throw error;
  }
};

/**
 * Get AI models status
 * @returns Promise with AI models availability status
 */
export const getAIModelsStatus = async (): Promise<AIModelsResponse> => {
  try {
    const response = await apiClient.get<AIModelsResponse>('/health/ai-models');
    return response.data;
  } catch (error) {
    throw error;
  }
};

/**
 * Get comprehensive health status of all system components
 * @returns Promise with detailed health information for all components
 */
export const getFullHealth = async (): Promise<FullHealthResponse> => {
  try {
    const response = await apiClient.get<FullHealthResponse>('/health/full');
    return response.data;
  } catch (error) {
    throw error;
  }
};

/**
 * Check if the backend API is reachable
 * @param timeout - Timeout in milliseconds
 * @returns Promise resolving to boolean indicating if API is reachable
 */
export const checkApiConnectivity = async (timeout = 5000): Promise<boolean> => {
  try {
    await apiClient.get('/health', {
      timeout,
      skipRetry: true, // Don't retry this request
    });
    return true;
  } catch (error) {
    return false;
  }
};
